generator/internal/printer.h

```diff
@@ -0,0 +1,56 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef GOOGLE_API_CODEGEN_INTERNAL_PRINTER_H_
+#define GOOGLE_API_CODEGEN_INTERNAL_PRINTER_H_
+
+#include <string>
+#include "src/google/protobuf/compiler/code_generator.h"
+#include "src/google/protobuf/io/printer.h"
+
+namespace pb = google::protobuf;
+
+namespace google {
+namespace api {
+namespace codegen {
+namespace internal {
+
+/**
+ * Wrapper around a google::protobuf::io::ZeroCopyOutputStream and a
+ * google::protobuf::io::Printer object so that they can be used for
+ * code generation. 
+ */
+class Printer {
+ public:
+  Printer(pb::compiler::GeneratorContext* generator_context,
+          std::string const& file_name)
+    : output_(generator_context->Open(file_name)),
+      printer_(new pb::io::Printer(output_.get(), '$', NULL)) {}
+
+  pb::io::Printer* operator->() & { return printer_.get(); }
+  pb::io::Printer const* operator->() const& { return printer_.get(); }
+
+  Printer(Printer const&) = delete;
+  Printer& operator=(Printer const&) = delete;
+
+ private:
+  std::unique_ptr<pb::io::ZeroCopyOutputStream> output_;
+  std::unique_ptr<pb::io::Printer> printer_;
```

Comment:
Same comment as from other review: why not make `printer_` a value attr instead of a pointer?

Reply:
Having `printer_` be a value means that its move constructor will be called as part of moving the outer `Printer` object. Having `printer_` be a pointer instead of a value means that we won't be calling the move constructor of `printer_` in that case, we will be moving the `unique_ptr` object. At least, that was my vague reasoning - does that make sense? Is there any benefit to having it be a value? (e.g. avoiding a call to `new`?)

Reply:
Okay, so `Printer` is intended to be movable, and we want to make move cheap.
The main benefits I can think of to using a value are cache friendliness and preventing a heap allocation.

Reply:
Reasonable points - I'll update it to be a value.